f-button@1.10.0 - f-button changes according to latest PIE designs

dandel10n opened the pull request:
## f-button@1.10.0
### Added
- `trailing-icon` and `leading-icon` named slots for ability to add an icon before or after button text
- `hasIcon` prop for spacing and icon colours css classes to be added
- validation for `hasIcon` prop to be able to add only `leading`/`trailing`/false

### Changed
- Button paddings to be aligned with the PIE designs
- XSmall button font size to be aligned with the PIE designs

## fozzie-components@3.49.0
### Updated
- Fozzie version to v5.0.0-beta.10 which has the latest pie design token variables.

Primary:
![Screenshot 2021-07-20 at 15 57 04](https://user-images.githubusercontent.com/19548183/126346661-579530ea-dddc-485b-b09d-1deb87294057.png)
![Screenshot 2021-07-20 at 15 57 13](https://user-images.githubusercontent.com/19548183/126346676-3059a5aa-943a-44b8-b181-90a7a68caafd.png)


Secondary:
![Screenshot 2021-07-20 at 15 58 46](https://user-images.githubusercontent.com/19548183/126346973-f8831505-1b32-4631-bb2e-76d62c9b3d83.png)
![Screenshot 2021-07-20 at 15 58 52](https://user-images.githubusercontent.com/19548183/126346982-fafc1f8d-b430-45c8-a5be-59a96dd6f327.png)


Outline:
![Screenshot 2021-07-20 at 15 59 36](https://user-images.githubusercontent.com/19548183/126347147-8b9d5422-ed36-41d3-854f-46c0d6ec4ba9.png)
![Screenshot 2021-07-20 at 15 59 43](https://user-images.githubusercontent.com/19548183/126347159-bfd32f09-709b-40ce-9b31-612f4cd21d6a.png)


Ghost:
![Screenshot 2021-07-20 at 16 00 38](https://user-images.githubusercontent.com/19548183/126347326-a28bfc0f-b2a1-4413-887b-2e64b5efc2a9.png)
![Screenshot 2021-07-20 at 16 00 32](https://user-images.githubusercontent.com/19548183/126347335-812a01aa-9c8b-4286-96a3-55ed748051cb.png)

dandel10n committed:
f-button@1.10.0 - f-button changes according to latest PIE designs

ashleynolan commented on packages/components/atoms/f-button/CHANGELOG.md:
```diff
@@ -4,6 +4,19 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+*July 20, 2021*
+
+### Added
+- `trailing-icon` and `leading-icon` named slots for ability to add an icon befor or after button text
```
```suggestion
- `trailing-icon` and `leading-icon` named slots for ability to add an icon before or after button text
```

ashleynolan commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -14,19 +14,37 @@
+
+            <span
+                :class="[
+                    (hasLeadingIcon ? $style['o-btn-icon'] : ''),
```
Would this work?

```suggestion
                    (hasLeadingIcon ? $style['o-btn-icon', 'o-btn-icon--leading'] : ''),
```

or if not, perhaps:

```suggestion
                    (hasLeadingIcon ? [$style['o-btn-icon'], $style['o-btn-icon--leading']] : ''),
```

ashleynolan commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -14,19 +14,37 @@
+
+            <span
+                :class="[
+                    (hasTrailingIcon ? $style['o-btn-icon'] : ''),
```
Same as above – guessing there is a syntax that will work here to apply both classes in one conditional?

ashleynolan commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -145,7 +171,7 @@ $btn-default-borderRadius              : $border-radius;
 $btn-default-font-size                 : 'body-l';
 $btn-default-weight                    : $font-weight-bold;
-$btn-default-padding                   : 11px 1.5em 13px;
+$btn-default-padding                   : 12px 24px;
```
In the past, we've need to have 1px higher on the bottom than the top to account for the way line-height is rendered. Is this still the case? (Would be worth comparing the before/after of the buttons to see the difference).

Good if we can get rid of this though – just wanted to check!

ashleynolan replied:
Also, I'd use `spacing(x3)` for the left/right padding (same below on line 221 and then `spacing(x2)` on line 226 and `spacing()` on line 230)

dandel10n replied:
I checked that with the equal spacing the button will be the same as on designs when with the current spacing it is not. I will change it to `spacing()` where applicable

ashleynolan commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -192,14 +218,17 @@ $btn-disabled-bgColor                  : $color-disabled-01;
+$btn-sizeLarge-padding                 : 14px 24px;
 $btn-sizeLarge-loading-color           : $color-content-interactive-primary;
 $btn-sizeLarge-loading-colorOpaque     : rgba($btn-sizeLarge-loading-color, $btn-default-loading-opacity);
+$btn-sizeLarge-iconHeight              : 23.5px;
```
Might be worth checking with design comparing with a real icon design, but it looks like the icon size itself in the button designs for large icons is `18 x 18` which is inside a `24 x 24` bounding box (so padded out by 6px).

Medium is then 18x18 (with 24x24 bounding box), small is `15x15` (bounding box is 20px) and small is 12x12 (bounding box of 16px).

ashleynolan commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -67,6 +85,14 @@ export default {
             type: Boolean,
             default: false
+        },
+        hasTrailingIcon: {
```
I'd condense this into one prop `hasIcon` that can take values `false`, `trailing` or `leading`. That way you never get into a weird spot where people could set true for both of these props and you have to decide which classes to apply.

ashleynolan reviewed (dismissed):
Few comments – looks good though otherwise.

dandel10n committed:
f-button@1.10.0:
- changed hasLeadingIcon and hasTrailingIcon to hasIcon
- added validation check for hasIcon prop
- changed icon sizes and added spacing around icons
- updated README

dandel10n committed:
f-button@1.10.0 - update changelog

dandel10n committed:
f-button@1.10.0 - changed some px spacing values to spacing()

dandel10n committed:
f-button@1.10.0 - added if statement around named icon slots

FayeCarter commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -14,19 +14,33 @@
+            <span
+                :class="[
+                    $style['o-btn-text'],
+                    (isLoading ? $style['o-btn-text--hidden'] : '')
```
Could the classes for these three spans be 
```
{ [$style['o-btn-icon'], $style['o-btn-icon--leading']] :   hasIcon === 'leading' }
{ [$style['o-btn-text--hidden']] : isLoading }
{ [$style['o-btn-icon'], $style['o-btn-icon--trailing']] : hasIcon === 'trailing' }
```

DamianMullins replied:
I agree, I think they should be. Not sure where the ternary approach came in but we should be using the Vue syntax for conditional classes IMO.

dandel10n replied:
It didn't allow me to use two classes in one list so I had to split them back in two separate checks

dandel10n committed:
f-button@1.10.0 - rewrite conditional css classes

DamianMullins commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -14,19 +14,36 @@
+            :class="$style['o-button-content']">
+
+            <span
+                v-if="hasIcon && hasIcon === 'leading'"
```
`hasIcon === 'leading'` should be split out into a computed property to keep logic out of the template and also to reduce duplication.

dandel10n replied:
Great suggesting, thanks, now everything looks much tidier!

DamianMullins commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -14,19 +14,36 @@
+            <span
+                v-if="hasIcon && hasIcon === 'trailing'"
+                :class="[{
+                    [$style['o-btn-icon']]: hasIcon === 'trailing',
```
Same as the last comment, `hasIcon === 'trailing'` should be split out into a computed property.

dandel10n committed:
f-button@1.10.0 - hasLeadingIcon and hasTrailingIcon computed props + tests

dandel10n committed:
Merge branch 'master' into f-button@1.10.0-leading-and-trailing-icon-support

dandel10n committed:
Merge branch 'master' into f-button@1.10.0-leading-and-trailing-icon-support

dandel10n committed:
Merge branch 'master' into f-button@1.10.0-leading-and-trailing-icon-support

ashleynolan commented on packages/components/atoms/f-button/src/components/_tests/Button.test.js:
```diff
@@ -120,6 +120,36 @@ describe('Button', () => {
+                // Arrange
+                const propsData = {
+                    hasIcon,
+                    isIcon: false
```
Just out of interest – what currently happens if `isIcon` is set to true and hasIcon is set to `leading` or `trailing`?

Does an error get thrown, or does the component attempt to append an icon to the icon button?

ashleynolan commented on packages/components/atoms/f-button/src/components/Button.vue:
```diff
@@ -135,6 +166,10 @@ export default {
             }
+
+            if (!VALID_BUTTON_ICON_POSITION.includes(this.hasIcon)) {
+                throw new TypeError(`hasIcon is set to "${this.hasIcon}", but it can only be one of the following buttonSizes: "${VALID_BUTTON_ICON_POSITION.join('", "')}"`);
```
Super small thing – but should probably say:

"but it can only be one of the following **values**"

rather than `buttonSizes`
